platforms/core-execution/workers/src/main/java/org/gradle/workers/internal/WorkersServices.java

```diff
@@ -65,23 +67,25 @@ public void registerProjectServices(ServiceRegistration registration) {
         registration.add(IsolatedClassloaderWorkerFactory.class);
     }
 
-    @SuppressWarnings("unused") // Used by reflection
-    private static class BuildSessionScopeServices {
+    private static class BuildSessionScopeServices implements ServiceProvider {
+        @Provides
         WorkerDirectoryProvider createWorkerDirectoryProvider(GradleUserHomeDirProvider gradleUserHomeDirProvider) {
             return new DefaultWorkerDirectoryProvider(gradleUserHomeDirProvider);
         }
 
+        @Provides
         ConditionalExecutionQueueFactory createConditionalExecutionQueueFactory(ExecutorFactory executorFactory, WorkerLimits workerLimits, WorkerLeaseService workerLeaseService) {
             return new DefaultConditionalExecutionQueueFactory(workerLimits, executorFactory, workerLeaseService);
         }
 
+        @Provides
         WorkerExecutionQueueFactory createWorkerExecutionQueueFactory(ConditionalExecutionQueueFactory conditionalExecutionQueueFactory) {
             return new WorkerExecutionQueueFactory(conditionalExecutionQueueFactory);
         }
     }
 
-    @SuppressWarnings("unused") // Used by reflection
-    private static class GradleUserHomeServices {
+    private static class GradleUserHomeServices implements ServiceProvider {
+        @Provides
         WorkerDaemonClientsManager createWorkerDaemonClientsManager(WorkerProcessFactory workerFactory,
                                                                     LoggingManagerInternal loggingManager,
                                                                     ListenerManager listenerManager,
```

Comment:
💭 This shows one of the advantages of the changes: no need to ignore unused warnings on service providers.